src/extension.ts

```diff
@@ -99,4 +102,27 @@ class ExtensionRuntime {
   public async dispose(): Promise<void> {
     await this.client?.stop();
   }
+
+  public async startClientAndWaitForVersion() {
+    this.client = this.client ?? await DafnyLanguageClient.create(this.context, this.statusOutput);
+    this.client.start();
+    await this.client.onReady();
+    this.languageServerVersion = await this.getLanguageServerVersionAfterStartup();
+  }
+
+  public async restart(): Promise<void> {
+    this.statusOutput.appendLine('Terminating Dafny...');
+    await this.dispose();
+    // The first subscription is the statusOutput and should not be disposed.
+    for(let i = 1; i < this.context.subscriptions.length; i++) {
+      this.context.subscriptions[i].dispose();
```

Comment:
Why do we need to do this? Shouldn't we only restart the LSP server?

Reply:
How about calling `client.stop()`, which tries to shutdown the language server and then disconnects from it, following by `client.start()` ?

Reply:
`this.dispose()` actually calls `client.stop()`
`this.initializeClient()` actually calls `client.start();`, but with proper logging.
So I also think it's the right way of doing.

Reply:
But.. right now I feel that we're cleaning up too. We're removing all traces of the Dafny extension and then adding them again, right? Doing `client.stop()` followed by `client.start()` would I think just restart the LSP server.

Reply:
Do you know why it's not working? Is that a bug in `client.stop()` then? The code is here: https://github.com/microsoft/vscode-languageserver-node/blob/main/client/src/common/client.ts#L1231

Which should lead to the client sending a shutdown request to the server here: https://github.com/microsoft/vscode-languageserver-node/blob/536cc154323d4d5d66c225a9ad5c5f259057f6f9/client/src/common/client.ts#L2069

Reply:
Thanks for sharing this code, I did not know where to find it exactly.

So it looks like I'm right: subscriptions are not part of the client, but of the context of the extension; however, many of them rely on the connection like this: https://github.com/microsoft/vscode-languageserver-node/blob/main/client/src/common/client.ts#L738

This `connection` is a snapshot that is [reset when the client restarts](https://github.com/microsoft/vscode-languageserver-node/blob/main/client/src/common/client.ts#L1444)
onNotification. Thus, it justifies that we need to dispose old notification disposables and recreate somes, otherwise, the extension would not work.

Reply:
Makes sense, thanks for the explanation!